nixos/modules/security/acme.nix

```diff
@@ -244,7 +267,16 @@ let
             ${data.postRun}
           fi
         '');
-      };
+
+      } // (optionalAttrs (data.webroot != null) {
+        # Lego always tries to create .well-known/acme-challenge, but if webroot is owned
+        # by the wrong user then it will crash and break cert renewal.
+        ExecStartPre = "+" + pkgs.writeShellScript "acme-${cert}-make-webroot" ''
+          mkdir -p '${data.webroot}/.well-known/acme-challenge'
+          cd '${data.webroot}'
+          chown 'acme:${data.group}' . .well-known .well-known/acme-challenge
```

Comment:
Yeah, that's a good idea. I'll give it a shot and see what happens

Reply:
Ok I ran into an issue implementing this. In the acme test suite, the configured webroot is `/var/lib/acme/.acme-challenge`, which is [the default](https://github.com/NixOS/nixpkgs/blob/9622a8981bc71964b6364bbf3115c65b2352875c/nixos/modules/services/web-servers/nginx/vhost-options.nix#L72). With the removal of tmpfiles, neither nginx nor the acme module create this folder if it does not exist. This is a problem because systemd will not create source folders for bind paths if they do not exist, generating the following error (which is misleading... it _is_ a source directory problem):
```
webserver # [   12.092725] systemd[872]: acme-a.example.test.service: Failed to set up mount namespacing: /run/systemd/unit-root/tmp/webroot: No such file or directory
```
Note that the preStart script I wrote uses `mkdir -p` which circumvents this issue. I also tried an inverse symlink, where I created a pseudo-webroot inside `/var/lib/acme/.lego/${cert}/webroot`, and then linked it to the service's `/tmp` folder, and then linked the acme-challenge folder from the pseudo-webroot into the real webroot. I encountered some 404's 
during renewal which I believe were due to permission errors, but I didn't investigate further as I realised this would not work if more than one cert shared the same webroot.

In terms of chowning webroot, I gave this a bit of thinking and came up with a better solution. If the webroot should be owned by acme, then it should have permission to create it too. This is true if it resides in /var/lib/acme, as it does by default. I moved this script into the main start script and removed the chown of webroot. I kept chown for the other 2 folders in case the configured group changes (e.g. switching web servers).